ext/neural_snow/ModelTools.jl

```diff
@@ -2,130 +2,481 @@ module ModelTools
-    dst::NamedTuple{(:up_bound, :low_bound, :output_pos)},
-    src::@NamedTuple{
-        up_bound::Tuple{},
-        low_bound::Tuple{},
-        output_pos::Tuple{},
-    };
-    filter = _ -> true,
-    cache = Base.IdSet(),
-)
-    return dst
+    paired_timeseries,
+    save_predictive_model_weights,
+    load_model_weights!,
+    freeze_fixed_layers!,
+    convert_model
+
+include("./WebTools.jl")
+
+""" 
+    MulLayer {T<:AbstractMatrix}
+Creates a custom Flux layer to reduce memory allocations and improve speed when no
+bias nor activation term is needed, and only a matrix multiplication layer.
+"""
+struct MulLayer{T <: AbstractMatrix{<:AbstractFloat}}
+    W::T
+end
+
+
+""" 
+   (layer::MulLayer)(x::AbstractArray{<:AbstractFloat})
+Defines the propogation behavior of the MulLayer and restricts the type
+"""
+function (layer::MulLayer)(x::AbstractArray{<:AbstractFloat})
+    layer.W * x
+end
+
+#Sets up/endows the layer type with Flux layer properties (gradient tracking, etc.)
+Flux.@layer MulLayer
+
+
+"""
+    connection_optimized(upperb, lowerb, pred, input)
+Generator function to create the memory/speed-optimized connection-function for 
+the neural network when provided with an upper-bound function `upperb(pred, input)`
+and lower-bound function `lowerb(pred, input)`, where `pred` refers
+to the set of predicted models by the predictive component of the model,
+and `input` refers to the inputs of the model. The connection function is generated with a call
+`connectfunc(pred, input) = connection_optimized(upperb, lowerb, pred, input)`
+for usage within the model, see `make_model()` or `make_model_paper()`.
+If one wishes to pass boundary functions taking additional arguments,
+the below code must be modified to expand the function's arguments
+and those passed to `upperb()` and `lowerb()`.
+# Arguments
+- `upperb::Function`: A function for calculating the upper-boundary value from `pred` and `input`
+- `lowerb::Function`: A function for calculating the lower-boundary value from `pred` and `input`
+- `pred::AbstractArray{<:AbstractFloat}`: The array of predictions from the predictive component.
+- `input::AbstractArray{<:AbstractFloat}`: The array of inputs sent to the predictive component for each prediction.
+    according with the predictive model structure, this array is N × K, where N is the number of features and K is the
+    number of predictions.
+"""
+function connection_optimized(
+    upperb::Function,
+    lowerb::Function,
+    pred::AbstractArray{T},
+    input::AbstractArray{T},
+)::AbstractArray{T} where {T <: AbstractFloat}
+    return [
+        upperb(pred, input)
+        lowerb(pred, input)
+        pred
+    ]
 end
 
+
+"""
+    paper_upper_bound(precip_idx, pred, input)
+Generator function to create the memory/speed-optimized upper-bound function
+from the paper when given which index of the input features pertains to the
+precipitation value. This also serves as an example of one way to write a
```

Comment:
Let's not call things "paper". There's many papers associated to ClimaLand. Assume your readers are general ClimaLand users